File: smtp/typescript/examples/basic-send.ts
/**
 * Basic Email Sending Example
 *
 * This example demonstrates the simplest way to send a plain text email
 * using the SMTP client.
 */

import {
  smtpClient,
  EmailBuilder,
  TlsMode,
  SmtpError,
} from '@integrations/smtp';

async function main() {
  // Create an SMTP client using the builder pattern
  // This example uses a typical SMTP configuration with STARTTLS on port 587
  const client = smtpClient()
    .host('smtp.example.com')
    .port(587)
    .credentials('user@example.com', 'your-password')
    .tlsMode(TlsMode.StartTls) // Use STARTTLS for secure connection
    .connectTimeout(10000) // 10 second connection timeout
    .commandTimeout(30000) // 30 second command timeout
    .build();

  try {
    // Build a simple text email using the EmailBuilder
    const email = new EmailBuilder()
      .from('sender@example.com')
      .to('recipient@example.com')
      .subject('Hello from SMTP Client')
      .text('This is a simple plain text email sent using the SMTP client library.')
      .build();

    console.log('Sending email...');

    // Send the email and get the result
    const result = await client.send(email);

    // The result contains information about the sent email
    console.log('Email sent successfully!');
    console.log('Message ID:', result.messageId);
    console.log('Accepted recipients:', result.acceptedRecipients.length);

    // You can also check if there were any rejected recipients
    if (result.rejectedRecipients.length > 0) {
      console.warn('Some recipients were rejected:');
      for (const rejected of result.rejectedRecipients) {
        console.warn(`  - ${rejected.recipient}: ${rejected.reason}`);
      }
    }

  } catch (error) {
    // Handle SMTP-specific errors
    if (error instanceof SmtpError) {
      console.error('SMTP Error:', error.message);
      console.error('Error kind:', error.kind);

      // Check if the error is retryable
      if (error.isRetryable) {
        console.error('This error might be temporary. Consider retrying.');
      }

      // Access enhanced status code if available
      if (error.enhancedCode) {
        console.error('Enhanced code:', error.enhancedCode.format());
      }
    } else {
      console.error('Unexpected error:', error);
    }
  } finally {
    // Always close the client to release resources
    await client.close();
    console.log('Client closed.');
  }
}

// Run the example
main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: smtp/rust/examples/basic_send.rs
//! Basic SMTP Email Sending Example
//!
//! This example demonstrates how to:
//! - Create an SMTP client with the builder pattern
//! - Construct a simple text email
//! - Send the email and handle the result

use smtp_integration::{SmtpClient, SmtpConfig, Email, EmailAddress, SmtpError};

#[tokio::main]
async fn main() -> Result<(), SmtpError> {
    // Configure the SMTP client
    // For this example, we'll use a local mail server or testing service
    let config = SmtpConfig::builder()
        .host("smtp.example.com")
        .port(587)
        .username("user@example.com")
        .password("your-password")
        .use_tls(true)
        .use_starttls(true)
        .timeout_seconds(30)
        .build()?;

    // Create the SMTP client
    println!("Creating SMTP client...");
    let client = SmtpClient::new(config)?;

    // Build a simple text email
    let email = Email::builder()
        .from(EmailAddress::new("sender@example.com", Some("Sender Name")))
        .to(EmailAddress::new("recipient@example.com", Some("Recipient Name")))
        .subject("Hello from Rust SMTP!")
        .text_body("This is a simple text email sent using the Rust SMTP integration.\n\nBest regards,\nThe SMTP Bot")
        .build()?;

    println!("Sending email...");
    println!("  From: {}", email.from.email);
    println!("  To: {}", email.to.iter().map(|a| a.email.as_str()).collect::<Vec<_>>().join(", "));
    println!("  Subject: {}", email.subject);

    // Send the email
    match client.send(&email).await {
        Ok(response) => {
            println!("Email sent successfully!");
            println!("  Message ID: {:?}", response.message_id);
            println!("  Status: {}", response.status);
        }
        Err(e) => {
            eprintln!("Failed to send email: {}", e);
            return Err(e);
        }
    }

    Ok(())
}
